File: Algorithms/Dijkstra's algorithm/Dijkstra's algorithm.py
import heapq

def dijkstra(graph, start):
    distances = {node: float('infinity') for node in graph}
    distances[start] = 0
    queue = [(0, start)]
    while queue:
        (current_distance, current_node) = heapq.heappop(queue)
        if current_distance > distances[current_node]:
            continue
        for neighbor, weight in graph[current_node].items():
            distance = current_distance + weight
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(queue, (distance, neighbor))
    return distances

def get_shortest_path(graph, start, end):
    distances = dijkstra(graph, start)
    return distances[end]

if __name__ == "__main__":
    graph = {
        'A': {'B': 7, 'C': 9, 'D': 14},
        'B': {'A': 7, 'C': 10, 'D': 15},
        'C': {'A': 9, 'B': 10, 'D': 11, 'E': 2},
        'D': {'A': 14, 'B': 15, 'C': 11, 'E': 6},
        'E': {'C': 2, 'D': 6}
    }
    start = 'A'
    end = 'E'
    shortest_path = get_shortest_path(graph, start, end)
    print(f"Shortest path from node {start} to node {end}: {shortest_path}")

File: java_practise/Dijkstra_algorithm/Dijkstra's algorithm.java
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.List;
import java.util.ArrayList;

public class Dijkstra {

    public static Map<Integer, Integer> dijkstra(List<List<Integer>> graph, int startNode) {
        int numNodes = graph.size();
        Map<Integer, Integer> shortestPaths = new HashMap<>();
        PriorityQueue<Node> queue = new PriorityQueue<>();

        for (int i = 0; i < numNodes; i++) {
            shortestPaths.put(i, Integer.MAX_VALUE);
        }
        shortestPaths.replace(startNode, 0);
        queue.offer(new Node(startNode, 0));

        while (!queue.isEmpty()) {
            Node currentNode = queue.poll();
            int currentDist = currentNode.dist;
            int currentIndex = currentNode.index;

            for (int i = 0; i < numNodes; i++) {
                if (graph.get(currentIndex).get(i) != 0) {
                    int newDist = currentDist + graph.get(currentIndex).get(i);
                    if (newDist < shortestPaths.get(i)) {
                        shortestPaths.replace(i, newDist);
                        queue.offer(new Node(i, newDist));
                    }
                }
            }
        }

        return shortestPaths;
    }

    public static int getShortestPath(List<List<Integer>> graph, int startNode, int endNode) {
        Map<Integer, Integer> shortestPaths = dijkstra(graph, startNode);
        return shortestPaths.get(endNode);
    }

    public static void main(String[] args) {
        List<List<Integer>> graph = new ArrayList<>();
        graph.add(List.of(0, 7, 9, 0, 0, 14));
        graph.add(List.of(7, 0, 10, 15, 0, 0));
        graph.add(List.of(9, 10, 0, 11, 0, 2));
        graph.add(List.of(0, 15, 11, 0, 6, 0));
        graph.add(List.of(0, 0, 0, 6, 0, 9));
        graph.add(List.of(14, 0, 2, 0, 9, 0));

        int startNode = 0;
        int endNode = 5;
        int shortestPath = getShortestPath(graph, startNode, endNode);

        System.out.println("Shortest path from node " + startNode + " to node " + endNode + ": " + shortestPath);
    }
}

class Node implements Comparable<Node> {
    int index;
    int dist;

    Node(int index, int dist) {
        this.index = index;
        this.dist = dist;
    }

    public int compareTo(Node other) {
        return Integer.compare(this.dist, other.dist);
    }
}
